// anv/index.ts
let anvInstance: any = null;
export function setInstance(instance: any) {
  if (!anvInstance) {
    anvInstance = instance;
  }
}

interface BaseFacet {
  facetId?: string;
};

// FIXME: Use better naming
export interface ProviderItem {
  title: string;
  cover?: string;
  sources: MediaSourceItem[];
}

export interface MediaSourceItem {
  type: "stream" | "mirror" | "mediasource";
  number?: string;
  title?: "string";
  resolver?: string;
  url: string;
  tiers?: string[];
  fileExtension: string;
}

export interface ProviderFacet extends BaseFacet {
  name: string;
  displayName?: string;
  description?: string;
  weight?: number;
  cacheSource?: boolean;
  delay?: number | ((ctx: {}, done: () => void) => void);
  resolvers: {
    mediaList?: string;
    mediaSource?: string;
  };
  hosts: string[];
  validUrl(url: string, list: boolean): boolean;
  tiers?: [string, string][];
  mediaList(metaMediaList: any): ProviderItem;
  mediaSource?(metaMediaSource: any, direct: boolean): MediaSourceItem[];
  search?(query: string): string[];
}

export interface MirrorResult {
  url: string;
  options?: {
    [option: string]: any;
  }
}

export interface MirrorFacet extends BaseFacet {
  name: string;
  displayName?: string;
  description?: string;
  weight?: number;
  cache?: boolean;
  delay?: number | ((ctx: {}, done: () => void) => void);
  resolver: string;
  hosts: string[];
  validUrl(url: string): boolean;
  tiers?: string[];
  media(metaMediaPage: string, tier: string): string | MirrorResult;
}

export interface GenericResolverFacet extends BaseFacet {
  name: string;
  description?: string;
  weight?: number;
  resolve(url: string, done: (err: string, resource: any) => void): void;
}

export interface StreamResolverFacet extends BaseFacet {
  name: string;
  description?: string;
  weight?: number;
  external?: boolean;
  resolve(url: string, out: string | NodeJS.WritableStream, info?: (data: {size: number; bytes: number}) => void, options?: StreamOptions): {
    stop(): void;
  }
}

export interface StreamOptions {
  [option: string]: any;
}

export type RegisterFuncFacets = "provider" | "mirror" | "genericresolver" | "streamresolver";

// FIXME: Refine these types
export interface RegisterFacetTypeMap {
  provider: ProviderFacet;
  mirror: MirrorFacet;
  genericresolver: GenericResolverFacet;
  streamresolver: StreamResolverFacet;
}

export function register<facet extends keyof RegisterFacetTypeMap>(facet: facet, facetOptions: RegisterFacetTypeMap[facet]): void {
  anvInstance.register(facet, facetOptions);
}

export function genericResolver(name: string, url: string, done: (data: any) => void) {
  anvInstance.genericResolver(name, url, done);
}
